src/Microsoft.DotNet.Interactive.Documents/Jupyter/Notebook.cs

```diff
@@ -0,0 +1,278 @@
+                                            "display_data" or "execute_result" =>
+                                                new InteractiveDocumentDisplayOutputElement(
+                                                    JsonSerializer.Deserialize<IDictionary<string, object>>(
+                                                        cellOutput.GetPropertyFromPath("data").GetRawText(), _serializerOptions)),
+
+                                            "stream" =>
+                                                new InteractiveDocumentTextOutputElement(
+                                                    GetTextAsSingleString(cellOutput.GetPropertyFromPath("text"))),
+
+                                            "error" =>
+                                                new InteractiveDocumentErrorOutputElement(
+                                                    cellOutput.GetPropertyFromPath("ename").GetString(),
+                                                    cellOutput.GetPropertyFromPath("evalue").GetString(),
+                                                    cellOutput.GetPropertyFromPath("traceback").EnumerateArray()
+                                                        .Select(s => s.GetString()).ToArray()),
+
+                                            _ => null
+                                        };
+                                    }
+
+                                    return null;
+                                })
+                                .Where(x => x is not null)
+                                .ToArray();
+                        }
+                        cells.Add(new InteractiveDocumentElement(cellLanguage, source, outputs));
+                        break;
+                    case "markdown":
+                        var markdown = GetTextAsSingleString(cell.GetPropertyFromPath("source"));
+                        cells.Add(new InteractiveDocumentElement("markdown", markdown));
+                        break;
+                }
+            }
+
+            return new InteractiveDocument(cells.ToArray());
+        }
+
+        public static InteractiveDocument Read(Stream stream, 
+            IDictionary<string, string> kernelLanguageAliases)
+        {
+            using var reader = new StreamReader(stream, Encoding);
+            var content = reader.ReadToEnd();
+            return Parse(content, kernelLanguageAliases);
+        }
+
+        public static async Task<InteractiveDocument> ReadAsync(Stream stream,
+            IDictionary<string, string> kernelLanguageAliases)
+        {
+            using var reader = new StreamReader(stream, Encoding);
+            var content = await reader.ReadToEndAsync();
+            return Parse(content, kernelLanguageAliases);
+        }
+
+        private static List<string> GetTextLines(JsonElement? jsonElement)
```

Comment:
Why allow null? I would think that would typically only result from a programming error.

Reply:
the element is fetched by trying to get it from the document, it can be null when not found

Reply:
When is it not found? Does this indicate a badly formed document? Should we be emitting a cell at all?

Reply:
We should emit the empty lines as we found a cell but the source element wasn't present.

Reply:
This is us being overly cautious in the scenario that the JSON is reporting that we're in a new cell, but there's no `"source"` property.  Rather than add a ton of `null` checks, it's easier to make these accept a nullable and default to empty strings/collections.